# Cleaning Kaggle Data

# Imports
import pandas as pd

In [1]:
# Imports
import pandas as pd

In [2]:
# Load in CSV
song_characteristics_raw = pd.read_csv('/Users/joseflemker/Documents/GitHub/3250-data-analytics-template/data/raw/Spotify_Song_Attributes.csv', encoding='utf-8')

In [3]:
# Examine first few rows and the data set as a whole
song_characteristics_raw.dtypes, print(song_characteristics_raw.head())

                                           trackName  \
0                                           "Honest"   
1  "In The Hall Of The Mountain King" from Peer G...   
2                                 #BrooklynBloodPop!   
3                                                $10   
4                         (I Just) Died In Your Arms   

                  artistName  msPlayed              genre  danceability  \
0               Nico Collins    191772                NaN         0.476   
1  London Symphony Orchestra   1806234  british orchestra         0.475   
2                       SyKo    145610         glitchcore         0.691   
3               Good Morning     25058   experimental pop         0.624   
4               Cutting Crew   5504949         album rock         0.625   

   energy   key  loudness  mode  speechiness  ...  liveness  valence    tempo  \
0   0.799   4.0    -4.939   0.0       0.2120  ...    0.2570    0.577  162.139   
1   0.130   7.0   -17.719   1.0       0.0510  ... 

(trackName            object
 artistName           object
 msPlayed              int64
 genre                object
 danceability        float64
 energy              float64
 key                 float64
 loudness            float64
 mode                float64
 speechiness         float64
 acousticness        float64
 instrumentalness    float64
 liveness            float64
 valence             float64
 tempo               float64
 type                 object
 id                   object
 uri                  object
 track_href           object
 analysis_url         object
 duration_ms         float64
 time_signature      float64
 dtype: object,
 None)

In [4]:
# Check for exact duplicates
duplicates = song_characteristics_raw.duplicated(keep=False)
print(f"duplicate records: {len(song_characteristics_raw[song_characteristics_raw.duplicated(keep='first')])} \n")


duplicate records: 5040 



In [5]:
# Examine some of the first values and first duplicates
duplicates_df = song_characteristics_raw[song_characteristics_raw.duplicated(keep=False)]
print(duplicates_df.head())

                                           trackName  \
0                                           "Honest"   
1  "In The Hall Of The Mountain King" from Peer G...   
2                                 #BrooklynBloodPop!   
3                                                $10   
4                         (I Just) Died In Your Arms   

                  artistName  msPlayed              genre  danceability  \
0               Nico Collins    191772                NaN         0.476   
1  London Symphony Orchestra   1806234  british orchestra         0.475   
2                       SyKo    145610         glitchcore         0.691   
3               Good Morning     25058   experimental pop         0.624   
4               Cutting Crew   5504949         album rock         0.625   

   energy   key  loudness  mode  speechiness  ...  liveness  valence    tempo  \
0   0.799   4.0    -4.939   0.0       0.2120  ...    0.2570    0.577  162.139   
1   0.130   7.0   -17.719   1.0       0.0510  ... 

In [6]:
# Now the original
print(song_characteristics_raw.head())

                                           trackName  \
0                                           "Honest"   
1  "In The Hall Of The Mountain King" from Peer G...   
2                                 #BrooklynBloodPop!   
3                                                $10   
4                         (I Just) Died In Your Arms   

                  artistName  msPlayed              genre  danceability  \
0               Nico Collins    191772                NaN         0.476   
1  London Symphony Orchestra   1806234  british orchestra         0.475   
2                       SyKo    145610         glitchcore         0.691   
3               Good Morning     25058   experimental pop         0.624   
4               Cutting Crew   5504949         album rock         0.625   

   energy   key  loudness  mode  speechiness  ...  liveness  valence    tempo  \
0   0.799   4.0    -4.939   0.0       0.2120  ...    0.2570    0.577  162.139   
1   0.130   7.0   -17.719   1.0       0.0510  ... 

Looks like all the data is duplicated twice so we will just drop the duplicates

In [7]:
# Drop Duplicates
song_characteristics_df = song_characteristics_raw.drop_duplicates()

In [8]:
# Double check there is no duplicates now
print(f"duplicate records: {len(song_characteristics_df[song_characteristics_df.duplicated(keep='first')])} \n")

duplicate records: 0 



In [9]:
# Now checking for duplicate song titles
duped_songs = song_characteristics_df[song_characteristics_df['trackName'].duplicated(keep=False)]
print(display(duped_songs))

trackName            artistName  msPlayed  \
42                    29           Chris James    138281   
43                    29       Run River North     36292   
76                  90mh  Albanos Xwris Xartia     76195   
77                  90mh              Trefuego   5128208   
161            Afterglow            Ed Sheeran     34603   
...                  ...                   ...       ...   
5007  You Broke Me First                Coopex   2013823   
5031          Youngblood   5 Seconds of Summer   1485347   
5032          Youngblood             Arc North   1009796   
5033             Younger            Jonas Blue   2560215   
5034             Younger                  Ruel   5272303   

                      genre  danceability  energy   key  loudness  mode  \
42    singer-songwriter pop         0.777   0.597   4.0    -5.125   1.0   
43         stomp and holler         0.439   0.903   7.0    -4.655   0.0   
76                      NaN           NaN     NaN   NaN       NaN   NaN   
77                      NaN           NaN     NaN   NaN       NaN   NaN   
161                     pop         0.641   0.324  11.0    -5.851   1.0   
...                     ...           ...     ...   ...       ...   ...   
5007           covertronica         0.661   0.796   1.0    -5.356   0.0   
5031               boy band         0.596   0.854   7.0    -5.114   0.0   
5032             slap house         0.465   0.891   7.0    -2.239   0.0   
5033                    pop         0.784   0.845   3.0    -2.793   1.0   
5034                  alt z         0.745   0.477  11.0    -7.706   0.0   

      speechiness  ...  liveness  valence    tempo            type  \
42         0.0496  ...    0.1070    0.900  133.854  audio_features   
43         0.0692  ...    0.2490    0.275  156.974  audio_features   
76            NaN  ...       NaN      NaN      NaN             NaN   
77            NaN  ...       NaN      NaN      NaN             NaN   
161        0.0299  ...    0.3280    0.273  110.184  audio_features   
...           ...  ...       ...      ...      ...             ...   
5007       0.1780  ...    0.0679    0.341  124.064  audio_features   
5031       0.4630  ...    0.1240    0.152  120.274  audio_features   
5032       0.3680  ...    0.2400    0.407  123.466  audio_features   
5033       0.0596  ...    0.0740    0.888  105.981  audio_features   
5034       0.0880  ...    0.1200    0.454  136.055  audio_features   

                          id                                   uri  \
42    0ahoVVrk5FRFzVF8Fp7Pzp  spotify:track:0ahoVVrk5FRFzVF8Fp7Pzp   
43    0y30gL6j3y3xDxK3sCOctP  spotify:track:0y30gL6j3y3xDxK3sCOctP   
76                       NaN                                   NaN   
77                       NaN                                   NaN   
161   0E4Y1XIbs8GrAT1YqVy6dq  spotify:track:0E4Y1XIbs8GrAT1YqVy6dq   
...                      ...                                   ...   
5007  0ISe0f9DauqBmQMnQcD2Ff  spotify:track:0ISe0f9DauqBmQMnQcD2Ff   
5031  2iUXsYOEPhVqEBwsqP70rE  spotify:track:2iUXsYOEPhVqEBwsqP70rE   
5032  6nLEcwqpVNOGRGtnV2pElc  spotify:track:6nLEcwqpVNOGRGtnV2pElc   
5033  4D7ERaKgv8NAeck8RzUtRz  spotify:track:4D7ERaKgv8NAeck8RzUtRz   
5034  2qXicQG06oT0ijKBznpgQv  spotify:track:2qXicQG06oT0ijKBznpgQv   

                                             track_href  \
42    https://api.spotify.com/v1/tracks/0ahoVVrk5FRF...   
43    https://api.spotify.com/v1/tracks/0y30gL6j3y3x...   
76                                                  NaN   
77                                                  NaN   
161   https://api.spotify.com/v1/tracks/0E4Y1XIbs8Gr...   
...                                                 ...   
5007  https://api.spotify.com/v1/tracks/0ISe0f9DauqB...   
5031  https://api.spotify.com/v1/tracks/2iUXsYOEPhVq...   
5032  https://api.spotify.com/v1/tracks/6nLEcwqpVNOG...   
5033  https://api.spotify.com/v1/tracks/4D7ERaKgv8NA...   
5034  https://api.spotify.com/v1/tracks/2qXicQG06oT0...   

  

None


Looks like most of the songs just happen to have the same name so we need to also pull artistName

In [10]:
# Now check with artist
duped_songs2 = song_characteristics_df[song_characteristics_df.duplicated(subset=['trackName', 'artistName'], keep=False)]
print(display(duped_songs2))

Empty DataFrame
Columns: [trackName, artistName, msPlayed, genre, danceability, energy, key, loudness, mode, speechiness, acousticness, instrumentalness, liveness, valence, tempo, type, id, uri, track_href, analysis_url, duration_ms, time_signature]
Index: []

[0 rows x 22 columns]

None


Duplicates are good now need to check for null values

In [11]:
# Print the total missing data
print(f'Total missing data: {song_characteristics_df.isna().sum().sum()}')
# Print missing data per column
print(song_characteristics_df.isna().sum())

Total missing data: 5700
trackName             0
artistName            0
msPlayed              0
genre               750
danceability        275
energy              275
key                 275
loudness            275
mode                275
speechiness         275
acousticness        275
instrumentalness    275
liveness            275
valence             275
tempo               275
type                275
id                  275
uri                 275
track_href          275
analysis_url        275
duration_ms         275
time_signature      275
dtype: int64


Seeing consisten 275 across all categories. Looks like we should drop all rows with danceability as null and verify the other columns go to 0

In [12]:
# Drop rows wheer danceability is null
song_characteristics_df = song_characteristics_df[song_characteristics_df['danceability'].notna()]

In [13]:
# Print missing data per column
print(song_characteristics_df.isna().sum())

trackName             0
artistName            0
msPlayed              0
genre               475
danceability          0
energy                0
key                   0
loudness              0
mode                  0
speechiness           0
acousticness          0
instrumentalness      0
liveness              0
valence               0
tempo                 0
type                  0
id                    0
uri                   0
track_href            0
analysis_url          0
duration_ms           0
time_signature        0
dtype: int64


This has cleared all but the genre column. Let's look at some of those values

In [14]:
# Make new df for null genres
null_genres_df = song_characteristics_df[song_characteristics_df['genre'].isna()]
print(null_genres_df)

                      trackName       artistName  msPlayed genre  \
0                      "Honest"     Nico Collins    191772   NaN   
15                          120            OQTAN    115125   NaN   
24                         1992          Bruises  20907439   NaN   
26                         2 AM       Marco Luka    172603   NaN   
31                          20s     Comfort Club     19957   NaN   
...                         ...              ...       ...   ...   
4964                Worst in Me     Unlike Pluto    236982   NaN   
4967             Would You Stay           Mindme   1139660   NaN   
4998                  Yesterday       Jammabands    155712   NaN   
5016           You Got Jebaited         Sordiway     22060   NaN   
5027  YouDeserveSomethingBetter  Charles Luciano   2270178   NaN   

      danceability  energy  key  loudness  mode  speechiness  ...  liveness  \
0            0.476   0.799  4.0    -4.939   0.0       0.2120  ...    0.2570   
15           0.806   0.94

It appears to be a random assortment of no genres. Because it is a releatively small amount and is necessary for analysis we will just remove the rows where genre is null.

In [15]:
# Drop rows wheer danceability is null
song_characteristics_df = song_characteristics_df[song_characteristics_df['genre'].notna()]

In [16]:
# Print missing data per column
print(song_characteristics_df.isna().sum())

trackName           0
artistName          0
msPlayed            0
genre               0
danceability        0
energy              0
key                 0
loudness            0
mode                0
speechiness         0
acousticness        0
instrumentalness    0
liveness            0
valence             0
tempo               0
type                0
id                  0
uri                 0
track_href          0
analysis_url        0
duration_ms         0
time_signature      0
dtype: int64


Now I need to ensure that the datatypes accurately represent the columns and we have only necessary columns included.

In [17]:
# Observe data types
print(song_characteristics_df.dtypes)

trackName            object
artistName           object
msPlayed              int64
genre                object
danceability        float64
energy              float64
key                 float64
loudness            float64
mode                float64
speechiness         float64
acousticness        float64
instrumentalness    float64
liveness            float64
valence             float64
tempo               float64
type                 object
id                   object
uri                  object
track_href           object
analysis_url         object
duration_ms         float64
time_signature      float64
dtype: object


In [18]:
# Look at total uniques for each column to ensure we get al the categorical variables we need
print(song_characteristics_df.nunique())

trackName           4113
artistName          1828
msPlayed            4240
genre                523
danceability         749
energy              1029
key                   12
loudness            3624
mode                   2
speechiness          963
acousticness        1822
instrumentalness    1792
liveness             951
valence             1099
tempo               3839
type                   1
id                  4261
uri                 4261
track_href          4261
analysis_url        4261
duration_ms         4076
time_signature         5
dtype: int64


In [19]:
# Based off those numbers and column descriptions choose categorical variables
print(f'Total genres: {len(song_characteristics_df['genre'].unique())}')
print(f' Total keys: {len(song_characteristics_df['key'].unique())}')
print(f' Total time signature: {len(song_characteristics_df['time_signature'].unique())}')
print(f' Total modes: {len(song_characteristics_df['mode'].unique())}')
print(song_characteristics_df['genre'].unique())
print(song_characteristics_df['key'].unique())
print(song_characteristics_df['time_signature'].unique())
print(song_characteristics_df['mode'].unique())

Total genres: 523
 Total keys: 12
 Total time signature: 5
 Total modes: 2
['british orchestra' 'glitchcore' 'experimental pop' 'album rock' 'alt z'
 'guitar case' 'cloud rap' 'dance pop' 'desi hip hop' 'lo-fi sleep'
 'contemporary country' 'bedroom r&b' 'singer-songwriter pop' 'la pop'
 'lo-fi chill' 'orchestral soundtrack' 'comic' 'alternative metal'
 'deep underground hip hop' 'pop' 'brooklyn drill' 'classical'
 'american orchestra' 'modern alternative rock' 'scandipop'
 'alabama indie' 'stomp and holler' 'anime score' 'dfw rap' 'punjabi pop'
 'folk-pop' 'acoustic pop' 'atl hip hop' 'filmi' 'art pop' 'japanese vgm'
 'sleep' 'electronica' 'australian pop' 'danish pop' 'melodic rap'
 'boy band' 'solipsynthm' 'ambient guitar' 'edm' 'bass house' 'show tunes'
 'chillwave' 'lo-fi study' 'post-teen pop' 'german soundtrack' 'hollywood'
 'gen z singer-songwriter' 'instrumental post-rock' 'baroque pop' 'r&b'
 'bedroom pop' 'dark r&b' 'chill r&b' 'canadian contemporary r&b'
 'aggressive phonk'

In [20]:
# Change columns to categorical
song_characteristics_df['genre_edited'] = song_characteristics_df['genre'].astype('category')
song_characteristics_df['key_edited'] = song_characteristics_df['key'].astype('category')
song_characteristics_df['time_signature_edited'] = song_characteristics_df['key'].astype('category')
song_characteristics_df['mode_edited'] = song_characteristics_df['mode'].astype('category')

In [21]:
# Show new dtypes
print(song_characteristics_df.dtypes)

trackName                  object
artistName                 object
msPlayed                    int64
genre                      object
danceability              float64
energy                    float64
key                       float64
loudness                  float64
mode                      float64
speechiness               float64
acousticness              float64
instrumentalness          float64
liveness                  float64
valence                   float64
tempo                     float64
type                       object
id                         object
uri                        object
track_href                 object
analysis_url               object
duration_ms               float64
time_signature            float64
genre_edited             category
key_edited               category
time_signature_edited    category
mode_edited              category
dtype: object


In [22]:
# Select and order the correct columns into cleaned dataframe
clean_df_columns = ['trackName', 'artistName', 'genre_edited', 'danceability', 'energy', 'key_edited', 'loudness', 'mode_edited', 'speechiness', 'acousticness', 'instrumentalness', 'liveness', 'valence', 'tempo', 'duration_ms', 'time_signature_edited']
song_characteristics_df_clean = song_characteristics_df[clean_df_columns]

In [23]:
# Rename to match data dictionairy
song_characteristics_df_clean = song_characteristics_df_clean.rename(columns={
                                     'trackName': 'track_name',
                                     'artistName': 'artist_name',
                                     'genre_edited': 'genre',
                                     'mode_edited': 'mode',
                                     'key_edited': 'key',
                                     'time_signature_edited': 'time_signature'
                                     })

In [24]:
# Save the dataframe to a csv
song_characteristics_df_clean.to_csv('/Users/joseflemker/Documents/GitHub/3250-data-analytics-template/data/finished/kaggle_finished_data.csv', encoding='utf-8', index=False)